File: DeskPet/AlarmManager.swift
import Foundation

protocol AlarmManagerDelegate: AnyObject {
    func alarmTriggered(_ alarm: AlarmItem, strength: ReminderStrength)
}

class AlarmManager {
    static let shared = AlarmManager()

    weak var delegate: AlarmManagerDelegate?
    private var checkTimer: Timer?
    private var firedToday: Set<UUID> = []
    private var lastCheckMinute: Int = -1
    private let settings = SettingsManager.shared
    private var snoozeTimers: [UUID: Timer] = [:]

    func start() {
        checkTimer?.invalidate()
        resetFiredIfNewDay()
        checkTimer = Timer.scheduledTimer(withTimeInterval: 15, repeats: true) { [weak self] _ in
            self?.tick()
        }
    }

    func rebuildAlarms() {
        snoozeTimers.values.forEach { $0.invalidate() }
        snoozeTimers.removeAll()
        firedToday.removeAll()
    }

    private func tick() {
        let now = Date()
        let cal = Calendar.current
        let hour = cal.component(.hour, from: now)
        let minute = cal.component(.minute, from: now)
        let encoded = hour * 60 + minute

        if encoded == 0 && lastCheckMinute != 0 {
            firedToday.removeAll()
        }
        lastCheckMinute = encoded

        for alarm in settings.alarms where alarm.enabled {
            let alarmEncoded = alarm.hour * 60 + alarm.minute
            guard encoded == alarmEncoded else { continue }
            guard !firedToday.contains(alarm.id) else { continue }
            guard let strength = settings.effectiveAlarmStrength(for: alarm) else { continue }

            firedToday.insert(alarm.id)

            if !alarm.repeatDaily {
                var updated = alarm
                updated.enabled = false
                settings.updateAlarm(updated)
            }

            DispatchQueue.main.async {
                self.delegate?.alarmTriggered(alarm, strength: strength)
            }
        }
    }

    private func resetFiredIfNewDay() {
        lastCheckMinute = -1
        firedToday.removeAll()
    }

    func snooze(_ alarm: AlarmItem) {
        snoozeTimers[alarm.id]?.invalidate()
        snoozeTimers[alarm.id] = Timer.scheduledTimer(withTimeInterval: 300, repeats: false) { [weak self] _ in
            guard let self = self else { return }
            self.snoozeTimers.removeValue(forKey: alarm.id)
            guard let strength = self.settings.effectiveAlarmStrength(for: alarm) else { return }
            DispatchQueue.main.async {
                self.delegate?.alarmTriggered(alarm, strength: strength)
            }
        }
    }
}

File: DeskPet.Windows/AlarmManager.cs
using System.Windows.Threading;

namespace DeskPet;

/// <summary>
/// Port of the macOS AlarmManager: a 15-second tick checks whether the current HH:MM
/// matches any enabled alarm. Each alarm fires at most once per day (tracked in
/// <c>_firedToday</c>, cleared at midnight). Non-repeating alarms auto-disable after
/// firing. Snooze re-fires once after 5 minutes.
/// </summary>
public sealed class AlarmManager
{
    public static readonly AlarmManager Shared = new();

    /// <summary>Raised when an alarm should be shown, with the resolved strength.</summary>
    public event Action<AlarmItem, ReminderStrength>? Triggered;

    private readonly DispatcherTimer _checkTimer = new() { Interval = TimeSpan.FromSeconds(15) };
    private readonly Dictionary<Guid, DispatcherTimer> _snoozeTimers = new();
    private readonly HashSet<Guid> _firedToday = new();
    private readonly SettingsManager _settings = SettingsManager.Shared;
    private int _lastCheckMinute = -1;

    private AlarmManager()
    {
        _checkTimer.Tick += (_, _) => Tick();
    }

    public void Start()
    {
        _checkTimer.Stop();
        _lastCheckMinute = -1;
        _firedToday.Clear();
        _checkTimer.Start();
    }

    /// <summary>Reset firing state (call after adding/editing alarms).</summary>
    public void RebuildAlarms()
    {
        foreach (var t in _snoozeTimers.Values) t.Stop();
        _snoozeTimers.Clear();
        _firedToday.Clear();
    }

    private void Tick()
    {
        var now = DateTime.Now;
        int encoded = now.Hour * 60 + now.Minute;

        // Clear the fired set once at the midnight minute.
        if (encoded == 0 && _lastCheckMinute != 0) _firedToday.Clear();
        _lastCheckMinute = encoded;

        // Snapshot so auto-disabling an alarm mid-loop doesn't mutate the collection we iterate.
        foreach (var alarm in _settings.Alarms.ToList())
        {
            if (!alarm.Enabled) continue;
            if (encoded != alarm.Hour * 60 + alarm.Minute) continue;
            if (_firedToday.Contains(alarm.Id)) continue;

            var strength = _settings.EffectiveAlarmStrength(alarm);
            if (strength is null) continue; // SuperDND

            _firedToday.Add(alarm.Id);

            if (!alarm.RepeatDaily)
            {
                alarm.Enabled = false;
                _settings.UpdateAlarm(alarm);
            }

            Triggered?.Invoke(alarm, strength.Value);
        }
    }

    /// <summary>Re-fire this alarm once, 5 minutes from now.</summary>
    public void Snooze(AlarmItem alarm)
    {
        if (_snoozeTimers.TryGetValue(alarm.Id, out var existing)) existing.Stop();

        var timer = new DispatcherTimer { Interval = TimeSpan.FromMinutes(5) };
        timer.Tick += (_, _) =>
        {
            timer.Stop();
            _snoozeTimers.Remove(alarm.Id);
            var strength = _settings.EffectiveAlarmStrength(alarm);
            if (strength is not null) Triggered?.Invoke(alarm, strength.Value);
        };
        _snoozeTimers[alarm.Id] = timer;
        timer.Start();
    }
}
